analyses/malawi/flooding_trigger/04_mwi_glofas_skill.md

```diff
@@ -0,0 +1,181 @@
+from scipy.stats import norm, pearsonr
+import numpy as np
+import pandas as pd
+import xarray as xr
+import xskillscore as xs
+from scipy.interpolate import interp1d
+from scipy.stats import rankdata
+import os
+from pathlib import Path
+import sys
+from datetime import timedelta
+from matplotlib.ticker import MaxNLocator, ScalarFormatter
+
+path_mod = f"{Path(os.path.dirname(os.path.abspath(''))).parents[1]}/"
+sys.path.append(path_mod)
+
+from src.indicators.flooding.config import Config
+from src.indicators.flooding.glofas import utils
+
+config = Config()
+mpl.rcParams['figure.dpi'] = 300
+
+PLOT_DIR = config.DATA_DIR / 'processed' / 'mwi' / 'plots' / 'flooding'
+PRIVATE_DIR = config.DATA_PRIVATE_DIR
+EXPLORE_DIR = PRIVATE_DIR / 'exploration' / 'mwi' / 'flooding'
+
+SAVE_FIG = True
+LEADTIMES = [x + 1 for x in range(10)]
+
+stations_adm2 = {
+    'G1724': 'Nsanje',
+    'G2001': 'Chikwawa'
+}
+COUNTRY_ISO3 = 'mwi'
+```
+
+Read in the processed GloFAS data and calculate the return periods. We can also plot the river discharge thresholds for each return period level.
+
+```python
+ds_glofas_reanalysis = utils.get_glofas_reanalysis(
+    country_iso3=COUNTRY_ISO3)
+ds_glofas_reforecast = utils.get_glofas_reforecast(
+    country_iso3 = COUNTRY_ISO3, leadtimes=LEADTIMES,
+    interp=False)
+df_return_period = utils.get_return_periods(ds_glofas_reanalysis)
+```
+
+```python
+rp_label = [str(int(x)) for x in df_return_period.index]
+rp_label[0] = '1.5'
+for code, station in stations_adm2.items():
+    fig, ax = plt.subplots()
+    ax.set_title(station)
+    rp = df_return_period[code]
+    ax.plot(rp_label, rp, 'o-', label=station)
+    ax.set_xlabel('Return period [years]')
+    ax.set_ylabel('River discharge [m$^3$ s$^{-1}$]')
+    ax.legend()
+    
+    if SAVE_FIG: plt.savefig(PLOT_DIR / f'{station}_return_periods.png')
+```
+
+Look into forecast skill by calculating the CRPS. We'll first do this by looking at all discharge values, and then recalculate looking specifically at extreme discharge values (eg. at the 3-year return period level). 
+
+```python
+def plot_crps(df_crps, title_suffix=None, ylog=False):
+    for code, station in stations_adm2.items():
+        fig, ax = plt.subplots()
+        crps = df_crps[code]
+        ax.plot(crps.index, crps, label=station)
+        ax.legend()
+        title = station
+        if title_suffix is not None:
+            title += title_suffix
+        ax.set_title(title)
+        ax.set_xlabel("Lead time [days]")
+        ax.set_ylabel("Normalized CRPS [% error]")
+        ax.xaxis.set_major_locator(MaxNLocator(integer=True))
+        ax.grid()
+        if ylog:
+            ax.set_yscale('log')
+            ax.yaxis.set_major_formatter(ScalarFormatter())
+```
+
+```python
+df_crps = utils.get_crps(ds_glofas_reanalysis, 
+                         ds_glofas_reforecast,
+                        normalization="mean")
+plot_crps(df_crps * 100, title_suffix=" -- all discharge values")
```

Comment:
It would be easier to compare the CRPS values if they were on the same figure (at least for the same station)

Reply:
Good point!